### PR TITLE
MGLLight autogenerate scripts

### PR DESCRIPTION
**Pull request by fabian-guerra**
Fixes https://github.com/mapbox/mapbox-gl-native/issues/9086

### CODE REVIEW

**Comment by fabian-guerra** [comment]
Work in progress ⚠️

File: platform/darwin/src/MGLLight.mm.ejs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+        auto <%- camelizeWithLeadingLowercase(property.name) -%>Value = mbglLight->get<%- camelize(property.name) -%>();
+        if (<%- camelizeWithLeadingLowercase(property.name) -%>Value.isUndefined()) {
+            _<%- camelizeWithLeadingLowercase(property.name) -%> = MGLStyleValueTransformer<<%- valueTransformerArguments(property).join(', ') %>>().toStyleValue(mbglLight->getDefault<%- camelize(property.name) -%>());
+        } else {
+            _<%- camelizeWithLeadingLowercase(property.name) -%> = MGLStyleValueTransformer<mbgl::style::Position, NSValue *>().toStyleValue(<%- camelizeWithLeadingLowercase(property.name) -%>Value);
+        }
+
+<% if (property.transition) { -%>        
+        _<%- camelizeWithLeadingLowercase(property.name) -%>Transiton = MGLTransitionFromOptions(mbglLight->get<%- camelize(property.name) -%>Transition());
```
Comment (frederoni) [inline platform/darwin/src/MGLLight.mm.ejs]: Nit: typo s/Transiton/Transition

File: platform/darwin/scripts/generate-style-code.js [hunk 2]
```diff
@@ -519,13 +522,29 @@ global.setSourceLayer = function() {
     return `_layer->setSourceLayer(sourceLayer.UTF8String);`
 };
 
+const lightProperties = Object.keys(spec[`light`]).reduce((memo, name) => {
```
Comment (1ec5) [inline platform/darwin/scripts/generate-style-code.js]: Nit: a backtick-delimited [template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) is only necessary for inserting `${variables}`. Otherwise, a simple `'string'` will do.

File: platform/darwin/scripts/style-spec-overrides-v8.json [hunk 3]
```diff
@@ -1,4 +1,5 @@
 {
+  "light-cocoa-doc": "An `MGLLight` object represents the light source for extruded geometries in `MGLStyle`.",
```
Comment (1ec5) [inline platform/darwin/scripts/style-spec-overrides-v8.json]: This is fine, but note that there’s no need to put every piece of Darwin-specific documentation in this file. The important thing is to put overrides here – things that would otherwise automatically say something different if we removed the entry from this file.

File: platform/darwin/src/MGLLight.h [hunk 4]
```diff
@@ -54,72 +63,131 @@ MGL_EXPORT
...
-@property (nonatomic) MGLTransition anchorTransition;
-
-
-/**
- Position of the light source relative to lit (extruded) geometries.
+ Position of the light source relative to lit (extruded) geometries, in [r
+ radial coordinate, a azimuthal angle, p polar angle] where r indicates the
+ distance from the center of the base of an object to its light, a indicates the
+ position of the light relative to 0° (0° when `light.anchor` is set to
+ `MGLLight.anchorViewport` corresponds to the top of the viewport, or 0° when
```
Comment (1ec5) [inline platform/darwin/src/MGLLight.h]: This should say `MGLLightAnchorViewport`, not `MGLLight.anchorViewport`. I wonder what happened here.

File: platform/darwin/src/MGLLight.h [hunk 5]
```diff
@@ -54,72 +63,131 @@ MGL_EXPORT
 @interface MGLLight : NSObject
 
 /**
- `anchor` Whether extruded geometries are lit relative to the map or viewport.
+ Whether extruded geometries are lit relative to the map or viewport.
+ 
+ The default value of this property is an `MGLStyleValue` object containing an
+ `NSValue` object containing `MGLAnchorViewport`.
```
Comment (1ec5) [inline platform/darwin/src/MGLLight.h]: This should say `MGLLightAnchorViewport`, not `MGLAnchorViewport`.

File: platform/darwin/src/MGLLight.h [hunk 6]
```diff
@@ -54,72 +63,131 @@ MGL_EXPORT
...
 /**
- Values describing animated transitions to `anchor` property.
- */
-@property (nonatomic) MGLTransition anchorTransition;
-
-
-/**
- Position of the light source relative to lit (extruded) geometries.
+ Position of the light source relative to lit (extruded) geometries, in [r
+ radial coordinate, a azimuthal angle, p polar angle] where r indicates the
```
Comment (1ec5) [inline platform/darwin/src/MGLLight.h]: This documentation comment is a good candidate for a Cocoa override, because it discusses an array syntax that isn’t relevant to Objective-C or Swift. The ideal documentation for this property would refer to `MGLLightAnchor` and let the documentation for that symbol describe the rest of the details.

File: platform/darwin/src/MGLLight.mm [hunk 7]
```diff
@@ -47,6 +51,50 @@ - (instancetype)initWithMBGLLight:(const mbgl::style::Light *)mbglLight
...
+
+        auto positionValue = mbglLight->getPosition();
+        if (positionValue.isUndefined()) {
+            _position = MGLStyleValueTransformer<mbgl::style::Position, NSValue *>().toStyleValue(mbglLight->getDefaultPosition());
+        } else {
+            _position = MGLStyleValueTransformer<mbgl::style::Position, NSValue *>().toStyleValue(positionValue);
+        }
+
+        
+        _positionTransiton = MGLTransitionFromOptions(mbglLight->getPositionTransition());
```
Comment (boundsj) [inline platform/darwin/src/MGLLight.mm]: s/Transiton/Transition in this file. This is causing the compiler errors.

**Comment by fabian-guerra** [comment]
@boundsj @1ec5 @frederoni could you please take a look.

File: include/mbgl/style/position.hpp [hunk 8]
```diff
@@ -9,7 +9,7 @@ namespace style {
 class Position {
 public:
     Position() = default;
-    Position(std::array<float, 3>& position_)
+    Position(const std::array<float, 3>& position_)
```
Comment (boundsj) [inline include/mbgl/style/position.hpp]: I thought I saw there might be a separate PR for this. Just checking you meant to leave this in.

Comment (fabian-guerra) [inline include/mbgl/style/position.hpp]: No, @jfirebaugh told me it was ok to leave it in this PR.

File: platform/darwin/scripts/style-spec-overrides-v8.json [hunk 9]
```diff
@@ -1,4 +1,10 @@
...
+  "light": {
+      "position": {
+        "doc": "Position of the `MGLLight` source relative to lit (extruded) geometries, in a `MGLSphericalPosition` struct [radial coordinate, azimuthal angle, polar angle] where radial indicates the distance from the center of the base of an object to its light, azimuthal indicates the position of the light relative to 0° (0° when `MGLLight.anchor` is set to `MGLLightAnchorViewport` corresponds to the top of the viewport, or 0° when `MGLLight.anchor` is set to `MGLLightAnchorMap` corresponds to due north, and degrees proceed clockwise), and polar indicates the height of the light (from 0°, directly above, to 180°, directly below)."
```
Comment (1ec5) [inline platform/darwin/scripts/style-spec-overrides-v8.json]: > [radial coordinate, azimuthal angle, polar angle]

This bit is unnecessary – it’s in the style specification because the JSON syntax literally requires an array with this ordering. Per https://github.com/mapbox/mapbox-gl-native/pull/9260#discussion_r121874710, everything from “where” onwards can be moved to the MGLLight documentation, since developers who want to learn how to use MGLLight might go there first.
